server/testing_service.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// NewTestingServer returns a new TestingServer for the Showcase API.
+func NewTestingServer(observerRegistry GrpcObserverRegistry) pb.TestingServer {
+	name := fmt.Sprintf("sessions/-")
+	defaultSession := NewSession(name, pb.Session_V1_LATEST, observerRegistry)
+	sessions := []sessionEntry{sessionEntry{session: defaultSession}}
+	keys := map[string]int{name: len(sessions) - 1}
+
+	s := &testingServerImpl{
+		token:            NewTokenGenerator(),
+		observerRegistry: observerRegistry,
+		keys:             keys,
+		sessions:         sessions,
+	}
+
+	return s
+}
+
+type sessionEntry struct {
+	session Session
+	deleted bool
+}
+
+type testingServerImpl struct {
+	uid              uniqID
+	token            TokenGenerator
+	observerRegistry GrpcObserverRegistry
+
+	mu       sync.Mutex
+	keys     map[string]int
+	sessions []sessionEntry
+}
+
+func (s *testingServerImpl) CreateSession(_ context.Context, req *pb.CreateSessionRequest) (*pb.Session, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	seshProto := req.GetSession()
+	id := s.uid.next()
+	name := fmt.Sprintf("sessions/%d", id)
+	sesh := NewSession(name, seshProto.GetVersion(), s.observerRegistry)
+
+	index := len(s.sessions)
+	s.sessions = append(s.sessions, sessionEntry{session: sesh})
+	s.keys[name] = index
+
+	return SessionProto(sesh), nil
+}
+
+func (s *testingServerImpl) GetSession(_ context.Context, req *pb.GetSessionRequest) (*pb.Session, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	name := req.GetName()
+	if i, ok := s.keys[name]; ok {
+		entry := s.sessions[i]
+		if !entry.deleted {
+			return SessionProto(entry.session), nil
+		}
+	}
+
+	return nil, status.Errorf(
+		codes.NotFound, "A session with name %s not found.",
+		name)
+}
+
+func (s *testingServerImpl) ListSessions(_ context.Context, in *pb.ListSessionsRequest) (*pb.ListSessionsResponse, error) {
+	start, err := s.token.GetIndex(in.GetPageToken())
+	if err != nil {
+		return nil, err
+	}
+
+	offset := 0
+	sessions := []*pb.Session{}
+	for _, entry := range s.sessions[start:] {
```

Comment:
I don't remember the `token.GetIndex()` logic at all, but would it be possible for me to supply a page token that resulted in a `start` index that's out of bounds here? It might be a concern if a gapic is jacking up the request payload, but maybe its not a real problem

Reply:
It's definitely possible. I tried to deter by making it hard to provide a valid page token that will result in a start that is out of bounds. This logic is everywhere so I'll make an issue to add a guard to protect from that.

Reply:
I filed: https://github.com/googleapis/gapic-showcase/issues/90